### PR TITLE
feat(grpo): add entropy regularization to GRPOTrainer

### LINKED ISSUE #5584: Add entropy regularization (e.g., `initial_entropy_coef`) to GRPOTrainer for improved exploration
### Feature request

I propose adding **explicit entropy regularization** to the GRPO training objective in TRL, with support for an **initial entropy coefficient and optional scheduling (decay over training steps)**.

Concretely, this would extend the current GRPO loss with an entropy term:

[
\mathcal{L} = \mathcal{L}_{GRPO} + \beta \cdot \mathcal{H}(\pi)
]

and expose user-facing parameters such as:

```python
GRPOConfig(
    initial_entropy_coef=0.02,
    final_entropy_coef=0.0,
    entropy_decay_steps=1000,
)
```

or a simpler version:

```python
GRPOConfig(
    entropy_coef=0.01,
)
```

This would provide a direct mechanism to control exploration at the **policy level**, rather than relying solely on sampling parameters (temperature, top-p).

**Related work / context:**

* Entropy regularization is standard in policy optimization methods such as PPO (e.g., Schulman et al., 2017)
* GRPO paper: https://arxiv.org/abs/2402.03300
* TRL implementation: https://github.com/huggingface/trl

### Motivation

Currently, exploration in GRPOTrainer is controlled indirectly via sampling parameters (e.g., temperature, top-p) and KL regularization.

In practice, this can lead to **very low exploration from the first training steps**, especially when:

* the base model has sharp logits
* KL penalty is non-negligible
* group samples become nearly identical

Observed issues include:

* very low diversity within GRPO groups
* near-zero reward variance across candidates
* weak or noisy learning signal
* early entropy collapse

Adjusting sampling parameters helps, but:

* it is indirect and unstable
* it does not affect the optimization objective itself
* it can still fail when KL constraints dominate

KL regularization toward a reference model does not reliably preserve entropy:

* if the reference policy is low-entropy → exploration collapses
* if KL is strong → deviations (including exploratory ones) are penalized

Adding entropy regularization would:

* encourage exploration directly in the objective
* improve diversity within groups → better relative ranking signal
* reduce sensitivity to sampling hyperparameters
* stabilize early training dynamics

This is a standard and well-understood technique in RL, and its absence in GRPOTrainer makes exploration fragile in practice.

### Your contribution

I’d be happy to contribute a PR implementing this feature.

The change appears relatively localized (loss computation + config), and I can:

* add entropy term to the GRPO loss
* implement a simple linear decay schedule
* expose config parameters in `GRPOConfig`
* include basic logging (entropy over time)

### PR DESCRIPTION
**Pull request by karthikchundi-commits**
# What does this PR do?

Adds explicit entropy regularization to the GRPO training objective to prevent entropy collapse and improve exploration during early training.

Introduces `entropy_coef`, `entropy_final_coef`, and `entropy_decay_steps` to `GRPOConfig`. When `entropy_coef > 0`, an entropy bonus is added to the loss:

```
L = L_GRPO - effective_coef * H(π)
```

The coefficient can optionally decay linearly from `entropy_coef` to `entropy_final_coef` over `entropy_decay_steps` training steps. Reuses the existing per-token entropy tensor already computed in `_get_per_token_logps_and_entropies()` — no additional forward pass overhead. Disabled by default (`entropy_coef=0.0`).

**Usage:**
```python
# Static coefficient
GRPOConfig(entropy_coef=0.01)

# With linear decay
GRPOConfig(entropy_coef=0.02, entropy_final_coef=0.0, entropy_decay_steps=1000)
```

Fixes #5584

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

@qgallouedec @lewtun @kashif — this implements the feature requested in #5584. Happy to add tests and docs if the approach looks good to you.

---

> [!NOTE]
> **Medium Risk**
> Changes the GRPO training objective by adding an optional entropy regularization term, which can affect training dynamics and metrics across runs. Risk is limited by defaulting the feature off (`entropy_coef=0.0`) and keeping the change localized to GRPO loss computation.
> 
> **Overview**
> **Adds entropy regularization to GRPO training (opt-in).** `GRPOConfig` now exposes `entropy_coef`, `entropy_final_coef`, and `entropy_decay_steps` to control an entropy bonus term.
> 
> **Updates `GRPOTrainer` loss computation** to subtract `effective_coef * mean_entropy` (with linear decay over `global_step` and accumulation-aware scaling) and logs the active `entropy_coef` as a metric when enabled.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7e4d4701aec82555173d27c54099b9c57ee12a6c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
See https://github.com/huggingface/trl/issues/5584#issuecomment-4268745144
